### PR TITLE
WIP: O11Y-2154: Implement MeterProvider

### PR DESCRIPTION
**Pull request by dustinlessard-wf**

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

**Comment by semveraudit-wf** [comment]
### Public API Changes

 Recommendation: **:warning: Minor version bump**


```diff
@@ line 6: package:opentelemetry/src/api/metrics/meterbuilder.dart @@
+  abstract class MeterBuilder
// Adding a class is a minor change.


```

```diff
@@ line 6: package:opentelemetry/src/api/metrics/meter.dart @@
+  abstract class Meter
// Adding a class is a minor change.


```

```diff
@@ line 13: package:opentelemetry/src/api/metrics/meter_provider.dart @@
+  abstract class MeterProvider
// Adding a class is a minor change.


```

Showing results for 21bbcd37f2f826fd67368dfd513bbf2379bb29f6
> Powered by [semver-audit-service](https://github.com/Workiva/semver-audit-service). Please report any problems by [filing an issue](https://github.com/Workiva/semver-audit-service/issues).
> Reported by the dart semver audit client 2.2.2
[Browse public API](https://semver-audit.wk-dev.wdesk.org/Workiva/opentelemetry-dart/21bbcd37f2f826fd67368dfd513bbf2379bb29f6).

**Comment by dustinlessard-wf** [comment]
@evanweible-wf @blakeroberts-wk , I'd like your input on a pattern I am questioning.

The OTEL specification organizes code into two main packages: API and SDK.  As you may know,
The API includes cross-cutting public interfaces used for instrumentation as well as default no-op implementations. Libraries are supposed to only depend on the API.
The SDK implements the API and is installed and managed by the application owner.  

Many languages implement the OTEL spec using 3 main constructs:
- Abstract classes (In the API)
- Noop classes (In the API)
- Concrete SDK classes

I propose that since Dart classes can be treated as interfaces, we forego the abstract classes so that we only have two main constructs:
- Classes that are in the API include the default noop implementation built in
- Classes that are in the SDK _implement_ the API classes

**Pros**
- simplifies the codebase

**Cons**
- deviates from typical code organization pattern

What do y'all think about this? Bad idea? Meh? Shrug? Awesome?

**Comment by evanweible-wf** [comment]
> I propose that since Dart classes can be treated as interfaces

FWIW, while this is true, the language team is actively discussing [proposals to introduce more control over types](https://github.com/dart-lang/language/issues/2242), which would include the ability to define an _actual_ interface.

Personally I would vote for having separate interfaces and no-op default implementations. Doing so gives you clearer separation of concerns, where each can be written and documented for its target audiences. The interface for implementers as well as consumers who want to know what the type means and what it should do, without needing to care about implementation-specific details; the no-ops can defer mostly to the interface, while making it clear to consumers that they indeed does nothing; and the concrete SDK implementations can explain how they implement the interface.

Combining the interfaces with the default no-op implementations also means that people could extend those classes without needing to implement every part of the interface (since they'd inherit the no-op behaviors by default). This creates a potential pitfall where they might implement only a subset of an interface, giving them an incomplete implementation.

**Comment by dustinlessard-wf** [comment]
Thanks gents! I'm going to rework this to separate the Noop implementations from the API classes.  I'll switch to using abstract classes that will force implementers to implement the required functions.

File: lib/sdk.dart [hunk 1]
```diff
@@ -35,3 +35,7 @@ export 'src/sdk/trace/span_processors/simple_processor.dart'
     show SimpleSpanProcessor;
 export 'src/sdk/trace/trace_state.dart' show TraceState;
 export 'src/sdk/trace/tracer_provider.dart' show TracerProviderBase;
+export 'src/sdk/metrics/meter.dart' show Meter;
```
Comment (dustinlessard-wf) [inline lib/sdk.dart:38]: should alphabetize the imports

File: lib/src/api/metrics/noop/noop_counter.dart [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import 'package:opentelemetry/api.dart';
+
+class NoopCounter<t> implements Counter {
```
Comment (dustinlessard-wf) [inline lib/src/api/metrics/noop/noop_counter.dart:3]: Comments?

File: lib/src/api/metrics/noop/noop_counter.dart [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+  @override
+  void add(dynamic value, {List<Attribute> attributes, Context context}) {}
```
Comment (dustinlessard-wf) [inline lib/src/api/metrics/noop/noop_counter.dart:5]: Need to figure this out.  Why can we do generics with the class, but have to use dynamic for the function? Why can't we do:
void add(t value, {List<Attribute> attributes, Context context}) {}

@evanweible-wf , is this just the way it has to be?

File: lib/src/api/metrics/noop/noop_meter.dart [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import 'package:opentelemetry/api.dart';
+
+class NoopMeter implements Meter {
```
Comment (dustinlessard-wf) [inline lib/src/api/metrics/noop/noop_meter.dart:3]: Should have a comment.

File: lib/src/sdk/metrics/counter.dart [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import 'package:opentelemetry/api.dart' as api;
+
+class Counter<t> implements api.Counter {
+  @override
+  void add(dynamic value,
+      {List<api.Attribute> attributes, api.Context context}) {
+    // TODO: implement add
```
Comment (dustinlessard-wf) [inline lib/src/sdk/metrics/counter.dart:7]: Should have a Github issue for any todos

File: lib/src/sdk/metrics/meter_provider.dart [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+      String schemaUrl = '',
+      Map<String, String> attributes = const {}}) {
+    if (instrumentationScopeName == null || instrumentationScopeName == '') {
+      instrumentationScopeName = '';
+      //throw and catch Argument error so that we capture a stacktrace,
+      //identifying the caller
+      try {
+        throw ArgumentError('Invalid Meter Name');
+      } catch (e, stacktrace) {
+        _logger.severe('Invalid Meter Name', e, stacktrace);
```
Comment (dustinlessard-wf) [inline lib/src/sdk/metrics/meter_provider.dart:21]: Should this be 'severe'?  It should probably be a warning instead.

File: lib/src/sdk/metrics/metric_options.dart [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import 'package:opentelemetry/api.dart' as api;
+
+class MetricOptions implements api.MetricOptions {
+  @override
+  void setDescription(String description) {
+    // TODO: implement setDescription
```
Comment (dustinlessard-wf) [inline lib/src/sdk/metrics/metric_options.dart:6]: Should have a Github issue for any todos

File: lib/src/sdk/metrics/metric_options.dart [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  }
+
+  @override
+  void setUnits(String units) {
+    // TODO: implement setUnits
```
Comment (dustinlessard-wf) [inline lib/src/sdk/metrics/metric_options.dart:11]: Should have a Github issue for any todos

File: test/unit/api/metrics/meter_provider_test.dart [hunk 9]
```diff
@@ -0,0 +1,43 @@
+@TestOn('vm')
+import 'package:opentelemetry/api.dart' as api;
+import 'package:test/test.dart';
+
+void main() {
+  group('MeterProvider:', () {
+    setUp(() {});
+
+    test('MeterProvider.get with name returns inert instance of Meter', () {
```
Comment (dustinlessard-wf) [inline test/unit/api/metrics/meter_provider_test.dart:9]: Nothing in these tests verify that the meter is inert. :(

Comment (dustinlessard-wf) [reply test/unit/api/metrics/meter_provider_test.dart:9]: Additionally, the object under test is the NoopMeterProvider, so perhaps I should change the name of the test.

File: test/unit/sdk/metrics/meter_provider_test.dart [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+      final meterProvider = sdk.MeterProvider();
+      final meterA = meterProvider.get(meterName,
+          instrumentationVersion: version, schemaUrl: url);
+      final meterB = meterProvider.get(meterName,
+          instrumentationVersion: version, schemaUrl: url);
+
+      expect(identical(meterA, meterB), true);
+    });
+
+    test(
```
Comment (dustinlessard-wf) [inline test/unit/sdk/metrics/meter_provider_test.dart:69]: I should add a test that verifies that getting by different name/etc returns a different meter instance.

File: test/unit/sdk/metrics/meter_provider_test.dart [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+          attributes: attributes);
+      final meterB = meterProvider.get(meterName,
+          instrumentationVersion: version,
+          schemaUrl: url,
+          attributes: attributes);
+
+      expect(identical(meterA, meterB), true);
+    });
+
+    // todo: imlpement test that verifies that changes to attributes apply to
```
Comment (dustinlessard-wf) [inline test/unit/sdk/metrics/meter_provider_test.dart:89]: spelling mistake

**Pull request closed without merging**